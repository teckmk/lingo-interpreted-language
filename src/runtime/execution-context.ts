export type StackFrame = {
  functionName: string
  filename: string
  position: {
    start: { line: number; column: number }
    end: { line: number; column: number }
  }
}

export class ExecutionContext {
  private callStack: StackFrame[]

  constructor() {
    this.callStack = []
  }

  public pushCallStack(frame: StackFrame) {
    this.callStack.push(frame)
  }

  public popCallStack() {
    this.callStack.pop()
  }

  public getCallStack(): StackFrame[] {
    return this.callStack.reverse()
  }

  static formatStackTrace(stack: StackFrame[], maxFrames = 10): string {
    const trimmedStack = stack.slice(0, maxFrames) // Limit frames to avoid excessive output
    let stackTrace = "Stack Trace:\n"
    for (const frame of trimmedStack) {
      // const shortFile = frame.filename.replace(/^.*[\\/]/, "") // Extract filename from path
      const shortFile = frame.filename
      stackTrace += `  at ${frame.functionName || "<anonymous>"} (${shortFile}:${frame.position.start.line}:${frame.position.start.column})\n`
    }

    if (stack.length > maxFrames) {
      stackTrace += `  ... (${stack.length - maxFrames} more frames)\n`
    }

    return stackTrace
  }
}
