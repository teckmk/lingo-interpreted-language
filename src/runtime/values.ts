import { Stmt } from "../frontend/2-ast"
import Environment from "./environment"

export type ValueType =
  | "null"
  | "number"
  | "string"
  | "boolean"
  | "object"
  | "nativefn"
  | "function"
  | "return"
  | "paramter"
  | "conditional"
  | "whileloop"
  | "array"
  | "dynamic"
  | "docs"

export interface RuntimeVal {
  type: ValueType
  returned: boolean
}

export interface NullVal extends RuntimeVal {
  type: "null"
  value: null
}
export interface BooleanVal extends RuntimeVal {
  type: "boolean"
  value: boolean
}

export interface NumberVal extends RuntimeVal {
  type: "number"
  value: number
}

export interface StringVal extends RuntimeVal {
  type: "string"
  value: string
}

export interface ObjectVal extends RuntimeVal {
  type: "object"
  properties: Map<string, RuntimeVal>
}

export type FunctionCall = (args: RuntimeVal[], env: Environment) => RuntimeVal

export interface NativeFnVal extends RuntimeVal {
  type: "nativefn"
  call: FunctionCall
}

export interface ParamVal extends RuntimeVal {
  type: "paramter"
  name: string
  valueType?: ValueType
  default: RuntimeVal // to assign a default value
}

export interface FunctionVal extends RuntimeVal {
  type: "function"
  name: string
  parameters: ParamVal[]
  declarationEnv: Environment
  body: Stmt[]
}

export interface ReturnVal extends RuntimeVal {
  type: "return"
  value: RuntimeVal
}

export interface ConditionalVal extends RuntimeVal {
  type: "conditional"
  check: RuntimeVal
  body: Stmt[]
  childChecks?: ConditionalVal[]
  else?: Stmt[]
}

export interface WhileLoopVal extends RuntimeVal {
  type: "whileloop"
  check: RuntimeVal
  body: Stmt[]
}

export interface ArrayVal extends RuntimeVal {
  type: "array"
  elements: RuntimeVal[]
}

export interface DocCommentVal extends RuntimeVal {
  type: "docs"
}
