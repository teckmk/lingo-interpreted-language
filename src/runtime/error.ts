import { ExecutionContext } from "./execution-context"

export class RuntimeError extends Error {
  context?: ExecutionContext

  constructor(context: ExecutionContext, message: string) {
    super(message)
    this.name = "RuntimeError"
    this.context = context
  }
}
