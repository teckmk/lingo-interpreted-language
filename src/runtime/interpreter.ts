import Environment from "./environment"
import { NumberVal, RuntimeVal, StringVal } from "./values"
import {
  ArrayLiteral,
  AssignmentExpr,
  BinaryExpr,
  CallExpr,
  DocComment,
  FunctionDeclaration,
  Identifier,
  IfElseStatement,
  MemberExpr,
  MultiVarDeclaration,
  NumericLiteral,
  ObjectLiteral,
  Program,
  ReturnStatement,
  Stmt,
  StringLiteral,
  VarDeclaration,
  WhileStatement,
} from "../frontend/2-ast"

import {
  eval_fn_declaration,
  eval_if_else_statement,
  eval_multi_var_declaration,
  eval_program,
  eval_return_statement,
  eval_var_declaration,
  eval_while_statement,
} from "./eval/statements"
import {
  eval_array_expr,
  eval_assignment,
  eval_binary_expr,
  eval_call_expr,
  eval_comment_expr,
  eval_identifier,
  eval_member_expr,
  eval_object_expr,
  eval_string_literal,
} from "./eval/expression"

export function evaluate(astNode: Stmt, env: Environment): RuntimeVal {
  switch (astNode.kind) {
    case "NumericLiteral":
      return {
        value: (astNode as NumericLiteral).value,
        type: "number",
      } as NumberVal
    case "StringLiteral":
      return eval_string_literal(astNode as StringLiteral, env) as StringVal
    case "BinaryExpr":
      return eval_binary_expr(astNode as BinaryExpr, env)
    case "AssignmentExpr":
      return eval_assignment(astNode as AssignmentExpr, env)
    case "Identifier":
      return eval_identifier(astNode as Identifier, env)
    case "ObjectLiteral":
      return eval_object_expr(astNode as ObjectLiteral, env)
    case "ArrayLiteral":
      return eval_array_expr(astNode as ArrayLiteral, env)
    case "CallExpr":
      return eval_call_expr(astNode as CallExpr, env)
    case "Program":
      return eval_program(astNode as Program, env)
    case "VarDeclaration":
      return eval_var_declaration(astNode as VarDeclaration, env)
    case "MultiVarDeclaration":
      return eval_multi_var_declaration(astNode as MultiVarDeclaration, env)
    case "FunctionDeclaration":
      return eval_fn_declaration(astNode as FunctionDeclaration, env)
    case "ReturnStatement":
      return eval_return_statement(astNode as ReturnStatement, env)
    case "IfElseStatement":
      return eval_if_else_statement(astNode as IfElseStatement, env)
    case "WhileStatement":
      return eval_while_statement(astNode as WhileStatement, env)
    case "MemberExpr":
      return eval_member_expr(astNode as MemberExpr, env)
    case "DocComment":
      return eval_comment_expr(astNode as DocComment, env)
    default:
      console.error("This AST can't be interpreted! For now atleast!", astNode)
      process.exit(0)
  }
}
