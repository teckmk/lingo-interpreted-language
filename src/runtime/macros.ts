import { BooleanVal, FunctionCall, NativeFnVal, NullVal, NumberVal, PlaceholderVal } from "./values"
import { PrimitiveTypeVal } from "./values.types"

export function MK_NUMBER(n = 0) {
  return { type: "number", value: n } as NumberVal
}

export function MK_NULL() {
  return { type: "null", value: null } as NullVal
}

export function MK_PLACEHOLDER() {
  return { type: "placeholder" } as PlaceholderVal
}

export function MK_BOOL(b: boolean) {
  return { type: "boolean", value: b } as BooleanVal
}

export function MK_NATIVE_FN(call: FunctionCall) {
  return { type: "nativefn", call } as NativeFnVal
}

export function MK_PRIM_TYPE(name: "number" | "bool" | "string" | "dynamic"): PrimitiveTypeVal {
  return {
    type: "type",
    typeName: name,
    typeKind: "primitive",
    primitiveType: name,
    returned: false, // To satify TS
  }
}
