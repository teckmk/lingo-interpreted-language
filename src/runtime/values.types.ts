import { RuntimeVal } from "./values"

export interface TypeVal extends RuntimeVal {
  type: "type"
  typeKind: string
  typeName?: string
  isNominal?: boolean // Track if this is a nominal type (created with 'type' keyword)
}

export interface PrimitiveTypeVal extends TypeVal {
  typeKind: "primitive"
  primitiveType: "string" | "number" | "bool" | "dynamic"
}

export interface StructTypeVal extends TypeVal {
  typeKind: "struct"
  members: Record<string, TypeVal>
  optional: Record<string, boolean> // Track which members are optional
}

export interface AliasTypeVal extends TypeVal {
  typeKind: "alias"
  aliasTo: TypeVal
}

export interface TypeParameterVal extends TypeVal {
  type: "type"
  typeKind: "typeParameter"
  name: string
  constraint?: TypeVal // i.e. T: number
}

export interface GenericTypeVal extends TypeVal {
  typeKind: "generic"
  baseType: TypeVal
  parameters: TypeParameterVal[]
}

export interface UnionTypeVal extends TypeVal {
  typeKind: "union"
  unionTypes: TypeVal[]
}

export interface ArrayTypeVal extends TypeVal {
  typeKind: "array"
  elementType: TypeVal
}
