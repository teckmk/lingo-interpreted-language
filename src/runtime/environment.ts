import { VarModifier } from "../frontend/ast"
import { ExecutionContext } from "./execution-context"
import { RuntimeError } from "./error"
import { RuntimeVal, ValueType } from "./values"

import functions from "./built-ins/functions"
import variables from "./built-ins/variables"

export default class Environment {
  private parent?: Environment
  private variables: Map<string, RuntimeVal>
  private constants: Set<string>
  private finals: Set<string>
  private executionContext: ExecutionContext

  constructor(executionContext: ExecutionContext, parentEnv?: Environment) {
    this.parent = parentEnv
    this.executionContext = executionContext
    this.variables = new Map()
    this.constants = new Set()
    this.finals = new Set()

    const global = Boolean(parentEnv) == false
    if (global) {
      // global variables
      for (const _var of variables) this.declareVar(_var.name, _var.value, _var.modifier)

      // global native functions
      for (const fn of functions) this.declareVar(fn.name, fn.value, fn.modifier)
    }
  }

  get context() {
    return this.executionContext
  }

  private assertType(varType: ValueType, value: RuntimeVal) {
    let valueWithType = value

    if (value.type == "dynamic") {
      valueWithType = { ...value, type: "dynamic" }
    } else if (value.type != varType && varType != "dynamic") {
      throw new RuntimeError(
        this.executionContext,
        `Can't assign a value of type ${value.type} to a variable of type ${varType}`,
      )
    }

    return valueWithType
  }

  public declareVar(
    varname: string,
    value: RuntimeVal,
    modifier: VarModifier,
    type?: ValueType,
  ): RuntimeVal {
    if (this.variables.has(varname))
      throw new RuntimeError(
        this.executionContext,
        `Cannot declare variable ${varname}. As it already is defined.`,
      )

    switch (modifier) {
      case "final":
        this.finals.add(varname)
        break
      case "constant":
        this.constants.add(varname)
        break
    }

    let valueWithType = value
    if (type) valueWithType = this.assertType(type, value)
    this.variables.set(varname, valueWithType)

    return valueWithType
  }

  public assignVar(varname: string, value: RuntimeVal): RuntimeVal {
    const env = this.resolve(varname)

    // Can't assign to a constant or final variable
    if (this.finals.has(varname))
      throw new RuntimeError(
        this.executionContext,
        `Cannot assign to a final variable "${varname}"`,
      )
    else if (this.constants.has(varname))
      throw new RuntimeError(
        this.executionContext,
        `Cannot assign to a constant variable "${varname}"`,
      )

    const prevVal = env.lookupVar(varname)

    const valueWithType = this.assertType(prevVal.type, value)

    env.variables.set(varname, valueWithType)

    return valueWithType
  }

  public resolve(varname: string): Environment {
    if (this.variables.has(varname)) return this

    if (this.parent == undefined)
      throw new RuntimeError(
        this.executionContext,
        `Cannot resolve ${varname} as it does not exist.`,
      )

    return this.parent.resolve(varname)
  }

  public lookupVar(varname: string): RuntimeVal {
    const env = this.resolve(varname)
    return env.variables.get(varname) as RuntimeVal
  }
}
