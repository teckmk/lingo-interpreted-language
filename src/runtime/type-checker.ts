import {
  AliasType,
  ArrayType,
  GenericType,
  PrimitiveType,
  TypeNode,
  UnionType,
} from "../frontend/ast"
import Environment from "./environment"
import { RuntimeError } from "./error"
import { ExecutionContext } from "./execution-context"
import { ObjectVal, RuntimeVal } from "./values"
import {
  AliasTypeVal,
  ArrayTypeVal,
  GenericTypeVal,
  PrimitiveTypeVal,
  StructTypeVal,
  TypeKind,
  TypeParameterVal,
  TypeVal,
  UnionTypeVal,
} from "./values.types"

function satisfiesConstraint(type: TypeVal, constraint: TypeVal): boolean {
  if (!constraint) {
    return true
  }

  const [isCompatible] = areTypesCompatible(type, constraint)
  return isCompatible
}

export function areTypesCompatible(
  sourceType: TypeVal,
  targetType: TypeVal,
  typeArgMap?: Map<string, TypeVal>, // Map of type parameter names to their concrete type arguments
): [boolean, TypeVal, TypeVal] {
  const bothNominal = sourceType.isNominal && targetType.isNominal
  // Check if types are nominally different
  if (bothNominal) {
    // Nominal types with different names are never compatible
    return [sourceType.typeName === targetType.typeName, sourceType, targetType]
  }

  // Handle type parameters with provided type arguments
  if (sourceType.typeKind === "typeParameter") {
    const sourceParam = sourceType as TypeParameterVal

    // If we have a concrete type for this parameter, use it instead
    if (typeArgMap && typeArgMap.has(sourceParam.name)) {
      const concreteType = typeArgMap.get(sourceParam.name)
      if (!concreteType) {
        throw new Error(`Type argument ${sourceParam.name} is not defined`)
      }
      return areTypesCompatible(concreteType, targetType, typeArgMap)
    }
  }

  if (targetType.typeKind === "typeParameter") {
    const targetParam = targetType as TypeParameterVal

    // If we have a concrete type for this parameter, use it instead
    if (typeArgMap && typeArgMap.has(targetParam.name)) {
      const concreteType = typeArgMap.get(targetParam.name)
      if (!concreteType) {
        throw new Error(`Type argument ${targetParam.name} is not defined`)
      }
      return areTypesCompatible(sourceType, concreteType, typeArgMap)
    }
  }

  // Dynamic type is compatible with anything
  if (
    (sourceType.typeKind === "primitive" &&
      (sourceType as PrimitiveTypeVal).primitiveType === "dynamic") ||
    (targetType.typeKind === "primitive" &&
      (targetType as PrimitiveTypeVal).primitiveType === "dynamic")
  ) {
    return [true, sourceType, targetType]
  }

  // Type parameter compatibility
  if (sourceType.typeKind === "typeParameter") {
    const sourceParam = sourceType as TypeParameterVal

    // If the target is also a type parameter, they are compatible if they have the same name
    // or if the source satisfies the target's constraint
    if (targetType.typeKind === "typeParameter") {
      const targetParam = targetType as TypeParameterVal

      // Same named parameters are compatible
      if (sourceParam.name === targetParam.name) {
        return [true, sourceType, targetType]
      }

      // If target has a constraint, source must satisfy it
      if (targetParam.constraint) {
        if (sourceParam.constraint) {
          // If source also has a constraint, check if source's constraint is compatible with target's
          const [isCompatible] = areTypesCompatible(
            sourceParam.constraint,
            targetParam.constraint,
            typeArgMap,
          )
          return [isCompatible, sourceType, targetType]
        } else {
          // If source has no constraint, it's not compatible with a constrained target
          return [false, sourceType, targetType]
        }
      }

      // Target has no constraint, so they're compatible
      return [true, sourceType, targetType]
    }

    // If source is a parameter and target is a type, check if source's constraint is compatible with target
    if (sourceParam.constraint) {
      const [isCompatible] = areTypesCompatible(sourceParam.constraint, targetType, typeArgMap)
      return [isCompatible, sourceType, targetType]
    }

    // Without constraints, can't determine compatibility
    return [false, sourceType, targetType]
  }

  // If target is a type parameter, check if source satisfies its constraint
  if (targetType.typeKind === "typeParameter") {
    const targetParam = targetType as TypeParameterVal

    if (targetParam.constraint) {
      const [isCompatible] = areTypesCompatible(sourceType, targetParam.constraint, typeArgMap)
      return [isCompatible, sourceType, targetType]
    }

    // No constraint means any type is acceptable
    return [true, sourceType, targetType]
  }

  // Same kind of primitives
  if (sourceType.typeKind === "primitive" && targetType.typeKind === "primitive") {
    return [
      (sourceType as PrimitiveTypeVal).primitiveType ===
        (targetType as PrimitiveTypeVal).primitiveType,
      sourceType,
      targetType,
    ]
  }

  // Check if source is compatible with any type in a union
  if (targetType.typeKind === "union") {
    for (const unionType of (targetType as UnionTypeVal).unionTypes) {
      const [isCompatible, resolvedSourceType, resolvedUnionType] = areTypesCompatible(
        sourceType,
        unionType,
        typeArgMap,
      )
      if (isCompatible) {
        return [true, resolvedSourceType, resolvedUnionType]
      }
    }
    return [false, sourceType, targetType]
  }

  // Alias types - check the underlying type
  if (sourceType.typeKind === "alias") {
    const [isCompatible, resolvedSourceType, resolvedTargetType] = areTypesCompatible(
      (sourceType as AliasTypeVal).aliasTo,
      targetType,
      typeArgMap,
    )
    return [isCompatible, resolvedSourceType, resolvedTargetType]
  }
  if (targetType.typeKind === "alias") {
    const [isCompatible, resolvedSourceType, resolvedTargetType] = areTypesCompatible(
      sourceType,
      (targetType as AliasTypeVal).aliasTo,
      typeArgMap,
    )
    return [isCompatible, resolvedSourceType, resolvedTargetType]
  }

  // Arrays - check element type compatibility
  if (sourceType.typeKind === "array" && targetType.typeKind === "array") {
    const [isCompatible, resolvedElementSourceType, resolvedElementTargetType] = areTypesCompatible(
      (sourceType as ArrayTypeVal).elementType,
      (targetType as ArrayTypeVal).elementType,
      typeArgMap,
    )

    if (!isCompatible) {
      return [false, sourceType, targetType]
    }

    // Create new array types with the resolved element types
    const resolvedSourceType: ArrayTypeVal = {
      ...(sourceType as ArrayTypeVal),
      elementType: resolvedElementSourceType,
    }

    const resolvedTargetType: ArrayTypeVal = {
      ...(targetType as ArrayTypeVal),
      elementType: resolvedElementTargetType,
    }

    return [true, resolvedSourceType, resolvedTargetType]
  }

  // Structs - check all required fields are compatible
  if (sourceType.typeKind === "struct" && targetType.typeKind === "struct") {
    const resolvedSourceMembers: Record<string, TypeVal> = {
      ...(sourceType as StructTypeVal).members,
    }
    const resolvedTargetMembers: Record<string, TypeVal> = {
      ...(targetType as StructTypeVal).members,
    }

    for (const [key, targetMemberType] of Object.entries((targetType as StructTypeVal).members)) {
      // Skip optional fields that don't exist in source
      if (
        (targetType as StructTypeVal).optional[key] &&
        !(sourceType as StructTypeVal).members[key]
      ) {
        continue
      }

      // Required field must exist and be compatible
      if (!(sourceType as StructTypeVal).members[key]) {
        return [false, sourceType, targetType]
      }

      const sourceMemberType = (sourceType as StructTypeVal).members[key]
      const [isCompatible, resolvedSourceMemberType, resolvedTargetMemberType] = areTypesCompatible(
        sourceMemberType,
        targetMemberType,
        typeArgMap,
      )

      if (!isCompatible) {
        return [false, sourceType, targetType]
      }

      resolvedSourceMembers[key] = resolvedSourceMemberType
      resolvedTargetMembers[key] = resolvedTargetMemberType
    }

    const resolvedSourceType: StructTypeVal = {
      ...(sourceType as StructTypeVal),
      members: resolvedSourceMembers,
    }

    const resolvedTargetType: StructTypeVal = {
      ...(targetType as StructTypeVal),
      members: resolvedTargetMembers,
    }

    return [true, resolvedSourceType, resolvedTargetType]
  }

  // Generic types - compare base types and parameters with constraint checking
  if (sourceType.typeKind === "generic" && targetType.typeKind === "generic") {
    if (sourceType.typeName !== targetType.typeName) {
      return [false, sourceType, targetType]
    }

    const sourceGeneric = sourceType as GenericTypeVal
    const targetGeneric = targetType as GenericTypeVal

    if (sourceGeneric.parameters.length !== targetGeneric.parameters.length) {
      return [false, sourceType, targetType]
    }

    // Create a new type arg map for this generic type comparison
    const newTypeArgMap = new Map(typeArgMap || [])

    // Process the parameters and build up type argument mapping
    for (let i = 0; i < sourceGeneric.parameters.length; i++) {
      const sourceParam = sourceGeneric.parameters[i]
      const targetParam = targetGeneric.parameters[i]

      // If we have concrete types for these parameters, add them to the map
      if (sourceParam.typeKind === "typeParameter" && targetParam.typeKind !== "typeParameter") {
        newTypeArgMap.set(sourceParam.name, targetParam)
      }

      // Check that parameters are compatible
      const [isCompatible] = areTypesCompatible(sourceParam, targetParam, newTypeArgMap)

      if (!isCompatible) {
        return [false, sourceType, targetType]
      }

      // Additional check: if target parameter has a constraint, source must satisfy it
      if (targetParam.constraint) {
        if (!satisfiesConstraint(sourceParam, targetParam.constraint)) {
          return [false, sourceType, targetType]
        }
      }
    }

    // If we got here, the generic types are compatible with the given type arguments
    return [true, sourceType, targetType]
  }

  return [false, sourceType, targetType]
}

// Helper function to create a type argument map from type parameters and arguments
export function createTypeArgMap(
  typeParameters: TypeParameterVal[],
  typeArguments: TypeVal[],
): Map<string, TypeVal> {
  const typeArgMap = new Map<string, TypeVal>()

  for (let i = 0; i < Math.min(typeParameters.length, typeArguments.length); i++) {
    const parameter = typeParameters[i]
    const { constraint } = parameter
    const typeArg = typeArguments[i]
    if (constraint && !satisfiesConstraint(typeArg, constraint)) {
      throw new Error(
        `Type argument ${getTypeName(typeArg)} does not satisfy constraint ${getTypeName(constraint)} of type parameter ${parameter.name}`,
      )
    }
    typeArgMap.set(parameter.name, typeArg)
  }

  return typeArgMap
}

// Function to instantiate a generic type with concrete type arguments
export function instantiateGenericType(
  genericType: GenericTypeVal,
  typeArguments: TypeVal[],
): TypeVal {
  // Create a map of type parameter names to concrete type arguments
  const typeArgMap = createTypeArgMap(genericType.parameters, typeArguments)

  // Create a deep copy of the base type
  const instantiatedType = JSON.parse(JSON.stringify(genericType.baseType))

  // Replace all type parameters with their concrete types
  return substituteTypeParameters(instantiatedType, typeArgMap)
}

// Helper function to substitute type parameters in a type with concrete types
function substituteTypeParameters(type: TypeVal, typeArgMap: Map<string, TypeVal>): TypeVal {
  if (type.typeKind === "primitive") {
    const paramName = (type as PrimitiveTypeVal).primitiveType
    if (typeArgMap.has(paramName)) {
      return typeArgMap.get(paramName) as TypeVal
    }
    return type
  }

  if (type.typeKind === "array") {
    const arrayType = type as ArrayTypeVal
    return {
      ...arrayType,
      elementType: substituteTypeParameters(arrayType.elementType, typeArgMap),
    } as ArrayTypeVal
  }

  if (type.typeKind === "alias") {
    const aliasType = type as AliasTypeVal
    return substituteTypeParameters(aliasType.aliasTo, typeArgMap)
  }

  if (type.typeKind === "struct") {
    const structType = type as StructTypeVal
    const members: Record<string, TypeVal> = {}

    for (const [key, memberType] of Object.entries(structType.members)) {
      members[key] = substituteTypeParameters(memberType, typeArgMap)
    }

    return {
      ...structType,
      members,
    } as StructTypeVal
  }

  if (type.typeKind === "generic") {
    const genericType = type as GenericTypeVal

    // Substitute parameters in the base type
    const substitutedBaseType = substituteTypeParameters(genericType.baseType, typeArgMap)

    return {
      ...genericType,
      baseType: substitutedBaseType,
    } as GenericTypeVal
  }

  // For other types, return as is
  return type
}

export function getRuntimeType(value: RuntimeVal): TypeVal {
  switch (value.type) {
    case "number":
      return {
        type: "type",
        typeKind: "primitive",
        primitiveType: "number",
      } as PrimitiveTypeVal

    case "string":
      return {
        type: "type",
        typeKind: "primitive",
        primitiveType: "string",
      } as PrimitiveTypeVal

    case "boolean":
      return {
        type: "type",
        typeKind: "primitive",
        primitiveType: "bool",
      } as PrimitiveTypeVal

    case "object":
      return {
        type: "type",
        typeKind: "struct",
        typeName: (value as ObjectVal).instanceOf,
        members: {},
        optional: {},
        isNominal: true,
      } as StructTypeVal

    // Handle other types as needed

    default:
      return {
        type: "type",
        typeKind: "primitive",
        primitiveType: "dynamic",
      } as PrimitiveTypeVal
  }
}

export function getTypeOfKind(typeVal: TypeVal, targetKind: TypeKind): TypeVal | undefined {
  // Base case: if the typeVal's typeKind matches the targetKind, return true
  if (typeVal.typeKind === targetKind) {
    return typeVal
  }

  // Handle different type kinds recursively, but skip struct members and array elements
  switch (typeVal.typeKind) {
    case "alias":
      return getTypeOfKind((typeVal as AliasTypeVal).aliasTo, targetKind)

    case "generic":
      return getTypeOfKind((typeVal as GenericTypeVal).baseType, targetKind)

    // case "union":
    //   return (typeVal as UnionTypeVal).unionTypes.some((unionType) =>
    //     isTypeOfKind(unionType, targetKind),
    //   )

    // case "typeParameter":
    // case "struct":
    // case "array":
    // case "primitive":
    //   // Do not check members or elements
    //   return false

    default:
      return
  }
}

export function getTypeName(type: TypeVal): string {
  switch (type.typeKind) {
    case "primitive":
      return (type as PrimitiveTypeVal).primitiveType

    case "struct":
      return type.typeName || "struct"

    case "array":
      return `${getTypeName((type as ArrayTypeVal).elementType)}[]`

    case "union":
      return (type as UnionTypeVal).unionTypes.map(getTypeName).join(" | ")

    case "generic": {
      const genericType = type as GenericTypeVal
      const baseTypeName = genericType.typeName || "generic"
      const paramNames = genericType.parameters.map((p) => p.name).join(", ")
      return `${baseTypeName}<${paramNames}>`
    }

    case "typeParameter":
      return (type as TypeParameterVal).name

    case "alias":
      return type.typeName || getTypeName((type as AliasTypeVal).aliasTo)

    default:
      return "unknown"
  }
}

export function isAliasOfType(type: TypeVal, targetTypeName: string, env: Environment): boolean {
  if (type.typeKind !== "alias") {
    return false
  }

  const aliasType = type as AliasTypeVal
  const aliasedType = aliasType.aliasTo as TypeVal

  // Direct match?
  if (aliasedType.typeName === targetTypeName) {
    return true
  }

  // Need to follow the alias chain
  if (aliasedType.typeKind === "alias") {
    return isAliasOfType(aliasedType, targetTypeName, env)
  } else if (aliasedType.typeKind === "primitive") {
    const primitiveType = aliasedType as PrimitiveTypeVal
    // Check if the primitive type points to our target
    return primitiveType.primitiveType === targetTypeName
  }

  return false
}

export function convertTypeNodeToTypeVal(
  typeNode: TypeNode,
  env: Environment,
  context: ExecutionContext,
): TypeVal {
  switch (typeNode.kind) {
    case "PrimitiveType":
      return {
        type: "type",
        typeKind: "primitive",
        primitiveType: (typeNode as PrimitiveType).name.value,
      } as PrimitiveTypeVal

    case "AliasType": {
      // Look up the actual type in the environment
      const typeName = typeNode.name?.value
      const isPredefinedPrimitive =
        typeName === "string" || typeName === "number" || typeName === "bool"

      if (typeName && !isPredefinedPrimitive) {
        const type = env.lookupType(typeName)
        if (!type) {
          throw new RuntimeError(context, `Unknown type '${typeName}'`)
        }
        return type as TypeVal
      }
      // If it's an inline alias type or predefined primitive, convert its actual type
      return convertTypeNodeToTypeVal((typeNode as AliasType).actualType, env, context)
    }

    case "StructType":
      // This would need to handle converting struct type nodes to StructTypeVal
      // Implementation depends on how you represent structs in your AST
      throw new RuntimeError(context, "Inline struct types not supported as type arguments yet")

    case "ArrayType":
      return {
        type: "type",
        typeKind: "array",
        elementType: convertTypeNodeToTypeVal((typeNode as ArrayType).elementType, env, context),
      } as ArrayTypeVal

    case "UnionType":
      return {
        type: "type",
        typeKind: "union",
        unionTypes: (typeNode as UnionType).types.map((t) =>
          convertTypeNodeToTypeVal(t, env, context),
        ),
      } as UnionTypeVal

    case "GenericType": {
      // For generic types, convert the base type and parameters
      const baseType = typeNode.name?.value ? env.lookupType(typeNode.name.value) : null
      if (!baseType) {
        throw new RuntimeError(
          context,
          `Unknown generic type '${typeNode.name?.value || "anonymous"}'`,
        )
      }

      // Convert type parameters
      const parameters = (typeNode as GenericType).parameters.map((param) => {
        return {
          type: "type",
          typeKind: "typeParameter",
          name: param.name.value,
          constraint: param.constraint
            ? convertTypeNodeToTypeVal(param.constraint, env, context)
            : undefined,
        } as TypeParameterVal
      })

      return {
        type: "type",
        typeKind: "generic",
        typeName: typeNode.name?.value,
        baseType: baseType as TypeVal,
        parameters,
      } as GenericTypeVal
    }
    default:
      throw new RuntimeError(context, `Unsupported type node kind: ${(typeNode as TypeNode).kind}`)
  }
}

export function getStructTypeFromGeneric(
  genericType: GenericTypeVal,
  typeArgs: TypeVal[],
  context: ExecutionContext,
): StructTypeVal {
  // Instantiate the generic type with concrete type arguments
  const instantiatedType = instantiateGenericType(genericType, typeArgs)

  // Get the struct type from the instantiated type
  const structType = getTypeOfKind(instantiatedType, "struct") as StructTypeVal

  if (!structType) {
    throw new RuntimeError(context, `Generic type does not resolve to a struct`)
  }

  return structType
}
