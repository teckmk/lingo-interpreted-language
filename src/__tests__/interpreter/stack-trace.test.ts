import { interpret } from "../../runtime/interpreter";

describe("Interpreter - Stack Trace", () => {
  it("should print stack trace", () => {
    const code = `
        fn foo() {
            bar()
        }

        fn bar() {
            baz()
        }

        fn baz() {
            const x = 5
            x = 10
        }

        foo()
    `;

    const stackTrace = interpret("test", code);

    expect(stackTrace).toMatchSnapshot();
  });
});
