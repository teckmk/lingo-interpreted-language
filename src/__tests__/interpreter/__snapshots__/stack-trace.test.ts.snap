// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interpreter - Stack Trace should print stack trace 1`] = `
"Cannot assign to a constant variable "x"
Stack Trace:
  at baz (main:7:13)
  at bar (main:3:13)
  at foo (main:15:9)
"
`;
